server/svix-server/src/worker.rs

```diff
@@ -291,6 +257,117 @@ fn bytes_to_string(bytes: bytes::Bytes) -> String {
+        .one(db)
+        .await?
+        .ok_or_else(|| {
+            Error::Generic(format!(
+                "Unexpected: message doesn't exist {}",
+                queue_task.clone().msg_id()
+            ))
+        })?;
+    let payload = msg.payload.as_ref().expect("Message payload is NULL");
+
+    let msg_clone = msg.clone();
+    let create_message_app = CreateMessageApp::layered_fetch(
+        cache.clone(),
+        db,
+        None,
+        msg_clone.app_id,
+        msg_clone.org_id,
+        Duration::from_secs(30),
+    )
+    .await?
+    .ok_or_else(|| Error::Generic(format!("Application doesn't exist: {}", &msg.app_id)))?;
+
+    let endpoints: Vec<CreateMessageEndpoint> = create_message_app
+        .endpoints
+        .iter()
+        .filter(|endpoint| match &queue_task {
+            QueueTask::MessageV1(task) => task.endpoint_id == endpoint.id,
+            QueueTask::MessageBatch(_) => true
+        })
+        .filter(|endpoint| {
+            return
+            // No disabled or deleted endpoints ever
+               !endpoint.disabled && !endpoint.deleted &&
+            (
+                // Manual attempt types go through regardless
+                queue_task.clone().trigger_type() == MessageAttemptTriggerType::Manual
+                || (
+                        // If an endpoint has event types and it matches ours, or has no event types
+                        endpoint
+                        .event_types_ids
+                        .as_ref()
+                        .map(|x| x.0.contains(&msg.event_type))
+                        .unwrap_or(true)
+                    &&
+                        // If an endpoint has no channels accept all messages, otherwise only if their channels overlap.
+                        // A message with no channels doesn't match an endpoint with channels.
+                        endpoint
+                        .channels
+                        .as_ref()
+                        .map(|x| !x.0.is_disjoint(msg.channels.as_ref().map(|x| &x.0).unwrap_or(&HashSet::new())))
+                        .unwrap_or(true)
+            ))})
+        .cloned()
+        .collect();
+
+    // TODO: remove this section once destinations are obsolete
+    let destinations = messagedestination::Entity::secure_find_by_msg(queue_task.clone().msg_id())
+        .all(db)
+        .await
+        .unwrap();
+    if destinations.is_empty() {
+        let destinations = endpoints
+            .iter()
+            .map(|endpoint| messagedestination::ActiveModel {
+                msg_id: Set(msg.id.clone()),
+                endp_id: Set(endpoint.id.clone()),
+                next_attempt: Set(Some(Utc::now().into())),
+                status: Set(MessageStatus::Sending),
+                ..Default::default()
+            });
+        messagedestination::Entity::insert_many(destinations)
+            .exec(db)
+            .await?;
```

Comment:
Using a fetch first is very wasteful (another roundtrip, load on the db, etc), and it's also not needed.
You only need to create the destinations if the type is Batch. Otherwise they already exist.